nvflare/app_common/aggregators/intime_accumulate_model_aggregator.py

```diff
@@ -74,29 +73,58 @@ def accept(self, shareable: Shareable, fl_ctx: FLContext) -> bool:
             return False
         assert isinstance(dxo, DXO)
 
-        if dxo.data_kind not in (DataKind.WEIGHT_DIFF, DataKind.WEIGHTS):
+        if dxo.data_kind not in (DataKind.WEIGHT_DIFF, DataKind.WEIGHTS, DataKind.COLLECTION):
             self.log_error(fl_ctx, "cannot handle data kind {}".format(dxo.data_kind))
             return False
 
-        if dxo.data_kind != self.expected_data_kind:
-            self.log_error(fl_ctx, "expect {} but got {}".format(self.expected_data_kind, dxo.data_kind))
-            return False
+        if dxo.data_kind == DataKind.COLLECTION:
+            for key in dxo.data.keys():
+                sub_dxo = dxo.data.get(key)
+                if sub_dxo.data_kind != self.expected_data_kind:
+                    self.log_error(
+                        fl_ctx,
+                        f"expect {self.expected_data_kind} but got {sub_dxo.data_kind} from {key} in DXO collection",
+                    )
+                    return False
+        else:
+            if dxo.data_kind != self.expected_data_kind:
+                self.log_error(fl_ctx, "expect {} but got {}".format(self.expected_data_kind, dxo.data_kind))
+                return False
 
         processed_algorithm = dxo.get_meta_prop(MetaKey.PROCESSED_ALGORITHM)
         if processed_algorithm is not None:
             self.log_error(fl_ctx, f"unable to accept shareable processed by {processed_algorithm}")
             return False
 
-        current_round = fl_ctx.get_prop(AppConstants.CURRENT_ROUND)
-        self.log_debug(fl_ctx, f"current_round: {current_round}")
-        client_name = shareable.get_peer_prop(key=ReservedKey.IDENTITY_NAME, default="?")
+        contributor_name = shareable.get_peer_prop(key=ReservedKey.IDENTITY_NAME, default="?")
         contribution_round = shareable.get_header(AppConstants.CONTRIBUTION_ROUND)
 
         rc = shareable.get_return_code()
         if rc and rc != ReturnCode.OK:
-            self.log_info(fl_ctx, f"Client {client_name} returned rc: {rc}. Disregarding contribution.")
+            self.log_info(fl_ctx, f"Contributor {contributor_name} returned rc: {rc}. Disregarding contribution.")
             return False
 
+        # accept current DXO or collection of DXOs
+        if self.expected_data_kind == DataKind.COLLECTION:
```

Comment:
This doesn't look right. The expected_data_kind is to make sure that each "weight carrying DXO" from each site must be of the same data kind.

So in case of COLLECTION, let's say there will be two "weight carrying DXOs", and each can carry a different type of weights. To support this, then you actually will need a dict of expected data kinds.

Suppose the key of the 1st DXO is "foo", and the key of the 2nd if "bar", you could have the dict of expected data kinds like:

expected_data_kind["foo"] = DataKind.WEIGHTS
expected_data_kind["bar"] = DataKind.WEIGHT_DIFF

You probably could organize the internal data structure of this aggregator based on these keys, as a dict of, say, AggrItem that contains a Helper object, aggr history, expected data kind, etc.

This would make your aggregator just a collection of aggregation processors.

Reply:
Good idea. Refatored.